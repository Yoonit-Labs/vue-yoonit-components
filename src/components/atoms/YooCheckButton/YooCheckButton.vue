<template lang="pug">
  div(
    :class="['yoo-checkbox', takeModifier]"
  )
    input(
      type="checkbox"
      :id="timeId"
      v-model="checkedField"
      @change="doCheck($event)"
    )
    label(
      :for="timeId"
      :class="['yoo-check','fas fa-check', ...takeModifierCheck]"
    )
      p(
        v-if="text !== ''"
      ) {{ text }}

      span(
        v-else
      ) {{ text }}

</template>

<script>
import PropsConfig from './Props.config'
export default {
  name: 'YooCheckButton',
  props: {
    text: {
      type: String,
      required: false,
      default: ''
    },
    size: {
      type: String,
      required: false,
      default: 'medium',
      validator: value => PropsConfig.size.options.includes(value)
    },
    initialValue: {
      type: Boolean,
      required: false,
      default: false
    },
    textPosition: {
      type: String,
      required: false,
      default: 'right',
      validator: value => PropsConfig.textPosition.options.includes(value)
    },
    switchCard: {
      type: Boolean,
      required: false,
      default: false
    }
  },
  data: () => ({
    checkedField: false,
    timeId: null
  }),
  beforeCreated () {},
  created () {},
  beforeMount () {},
  mounted () {
    this.checkedField = this.initialValue
    this.timeId = `ycid-${Math.random().toString().substr(-7)}`
  },
  updated () {},
  beforeDestroy () {},
  destroyed () {},
  components: {},
  computed: {
    takeModifierCheck () {
      const block = 'yoo-check'
      const classList = []
      switch (this.size) {
        case 'small':
          classList
            .push(`${block}--${this.size}`)
          break
        case 'medium':
          classList
            .push(`${block}--${this.size}`)
          break
        case 'large':
          classList
            .push(`${block}--${this.size}`)
          break
      }
      if (this.checkedField) {
        classList
          .push(`${block}--checked`)
      }
      if (this.switchCard) {
        classList
          .push(`${block}--switch-card`)
      }

      classList
        .push(`${block}--text-${this.textPosition}`)

      return classList
    },
    takeModifier () {
      const block = 'yoo-checkbox'
      const classList = []
      if (this.switchCard) {
        classList
          .push(`${block}--switch-card`)
      }

      return classList
    }
  },
  methods: {
    doCheck (e) {
      this.$emit('response', this.checkedField)
    }
  },
  filters: {},
  watch: {}
}
</script>

<style src="./YooCheckButton.sass" lang="sass" scoped></style>
