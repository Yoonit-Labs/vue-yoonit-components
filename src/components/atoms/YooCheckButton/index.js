import YooCheckButton from './YooCheckButton.vue'
export default YooCheckButton
