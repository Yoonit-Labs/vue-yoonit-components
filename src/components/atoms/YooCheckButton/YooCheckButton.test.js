import { shallowMount } from '@vue/test-utils'
import YooCheckButton from '@/components/atoms/YooCheckButton/YooCheckButton.vue'
import PropsConfig from './Props.config'

const classBlock = 'yoo-check'
const SlotText = 'Default Slot Text'

const mountComponent = () => {
  return shallowMount(YooCheckButton, {
    slots: { default: SlotText }
  })
}

describe('YooCheckButton Component', () => {
  let wrapper
  beforeEach(() => {
    jest.spyOn(global.Math, 'random').mockReturnValue(0.123456789)
    wrapper = mountComponent()
  })

  afterEach(() => {
    jest.spyOn(global.Math, 'random').mockRestore()
  })

  it('Matches Snapshot', async () => {
    expect(wrapper).toMatchSnapshot()
  })

  it('Loads the Component HTML', () => {
    expect(wrapper.classes('yoo-checkbox')).toBe(true)
  })

  describe('Props', () => {
    describe('size', () => {
      it('Has a valid default value', () => {
        expect(PropsConfig.size.options.includes(YooCheckButton.props.size.default)).toBe(true)
      })
      PropsConfig.size.options.forEach(size => {
        it(`Includes size class: .${classBlock}--${size}`, async () => {
          await wrapper.setProps({ size })
          expect(wrapper.find(`.${classBlock}--${size}`).exists()).toBe(true)
        })
      })
    })

    describe('textPosition', () => {
      it('Has a valid default value', () => {
        expect(PropsConfig.textPosition.options.includes(YooCheckButton.props.textPosition.default)).toBe(true)
      })
      PropsConfig.textPosition.options.forEach(textPosition => {
        it(`Includes textPosition class: .${classBlock}--text-${textPosition}`, async () => {
          await wrapper.setProps({ textPosition })
          expect(wrapper.find(`.${classBlock}--text-${textPosition}`).exists()).toBe(true)
        })
      })
    })

    describe('initialValue', () => {
      it('Includes initialValue class: .yoo-check--checked when initialValue', async () => {
        wrapper = shallowMount(YooCheckButton, { propsData: { initialValue: true } })
        await wrapper.vm.$nextTick()
        expect(wrapper.find(`.${classBlock}--checked`).exists()).toBe(true)
      })
      it('Does not include initialValue class: .yoo-check--checked when not initialValue', async () => {
        await wrapper.setProps({ initialValue: false })
        expect(wrapper.find(`.${classBlock}--checked`).exists()).toBe(false)
      })
    })

    describe('switchCard', () => {
      it('Has a valid default value', () => {
        expect(YooCheckButton.props.switchCard.default).toBe(false)
      })
      it(`Includes switch-card class: .${classBlock}--switch-card when switchCard`, async () => {
        await wrapper.setProps({ switchCard: true })
        expect(wrapper.find(`.${classBlock}--switch-card`).exists()).toBe(true)
      })
      it(`Does not include switch-card class: .${classBlock}--switch-card when not switchCard`, async () => {
        await wrapper.setProps({ switchCard: false })
        expect(wrapper.find(`.${classBlock}--switch-card`).exists()).toBe(false)
      })
    })
  }) // describe Props

  describe('Events', () => {
    describe('Click', () => {
      it('Emits change Event', async () => {
        wrapper.find('input').trigger('click')
        expect(wrapper.emitted()).toHaveProperty('response')
      })
    })
  }) // describe Events
})
