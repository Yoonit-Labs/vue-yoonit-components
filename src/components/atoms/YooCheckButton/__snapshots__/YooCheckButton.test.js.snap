// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`YooCheckButton Component Matches Snapshot 1`] = `<div class="yoo-checkbox"><input type="checkbox" id="custom-time-id"><label class="yoo-check fas fa-check yoo-check--medium yoo-check--text-right" for="custom-time-id"><span></span></label></div>`;
